### PR TITLE
Add tests for picai transforms and other test fixes

### PR DESCRIPTION
**Pull request by jewelltaylor**
# PR Type
[Feature | Fix | Documentation | **Other** ]

# Short Description

Clickup Ticket(s): [Add Picai Tranforms Tests](https://app.clickup.com/t/8688q3ah7)

In #155, I added a preprocessing transformation pipeline for PICAI. In this PR, I simply have added tests for those transforms and also fixed a small issue in loss tests from previous PRs when using GPU.

# Tests Added

- Tests for PICAI transforms
- Updated Loss tests

### CODE REVIEW

File: research/picai/preprocessing_transforms.py [hunk 1]
```diff
@@ -33,7 +33,7 @@ def resample_img(
     original_size = image.GetSize()
 
     # convert PICAI z, y, x (Depth x Height x Width) convention to SimpleITK's convention x, y, z
-    out_spacing = list(reversed(spacing))
+    out_spacing = list(reversed(list(spacing)))
```
Comment (emersodb) [inline research/picai/preprocessing_transforms.py:36]: I know reversed was my suggestion. However, now that we've moved from sequences to tuples, maybe we just do the following?
```
out_spacing = (spacing[2], spacing[1], spacing[0])
```

File: research/picai/preprocessing_transforms.py [hunk 2]
```diff
@@ -157,13 +157,14 @@ def crop_or_pad(
...
     shape, size = input_verification_crop_or_pad(image, size, physical_size)
 
-    # set identity operations for cropping and padding
-    rank = len(size)
-    padding = [[0, 0]] * rank
-    slicer = [slice(None)] * rank
+    # Since the subarrays are being set in the below for loop
+    # We have to ensure that they are seperate lists
+    # and not the same reference (ie [[0, 0]] * rank)
+    padding = [[0, 0] for _ in range(len(size))]
```
Comment (emersodb) [inline research/picai/preprocessing_transforms.py:163]: Ah interesting. I didn't realize this tied them through reference like that. It doesn't work that way for things like `ints`. Good to know for the future.

File: research/picai/preprocessing_transforms.py [hunk 3]
```diff
@@ -177,11 +178,12 @@ def crop_or_pad(
             idx_end = idx_start + size[i]
             slicer[i] = slice(idx_start, idx_end)
 
+    new_image = image[tuple(slicer)]
+
     # crop and/or pad image
-    if isinstance(image, sitk.Image):
+    if any(pad[0] > 0 or pad[1] > 0 for pad in padding):
```
Comment (emersodb) [inline research/picai/preprocessing_transforms.py:184]: Any reason we're "dropping" support for non-sitk.Image types?

File: research/picai/preprocessing_transforms.py (hunk 2 above)
Comment (jewelltaylor) [reply research/picai/preprocessing_transforms.py:163]: Yeah I did not realize it either, I think it is only for mutable types

File: tests/losses/test_contrastive_loss.py [hunk 4]
```diff
@@ -18,21 +18,21 @@ def test_contrastive_loss() -> None:
         features=local_features.reshape(len(local_features), -1),
         positive_pairs=global_features.reshape(1, len(global_features), -1),
         negative_pairs=previous_local_features.reshape(1, len(previous_local_features), -1),
-    )
+    ).cpu()
```
Comment (emersodb) [inline tests/losses/test_contrastive_loss.py:21]: I'm not sure if you think this is better or not, but `.item()` automatically moves the value to cpu and produces a plain python float.

File: research/picai/preprocessing_transforms.py (hunk 3 above)
Comment (jewelltaylor) [reply research/picai/preprocessing_transforms.py:184]: I definitely want to support a wider variety of image types and for some its pretty trivial, but at this point just to ensure everything matches how its currently typed which doesn't include numpy arrays

File: tests/losses/test_contrastive_loss.py (hunk 4 above)
Comment (jewelltaylor) [reply tests/losses/test_contrastive_loss.py:21]: yeah sure I can do that

File: research/picai/preprocessing_transforms.py (hunk 3 above)
Comment (emersodb) [reply research/picai/preprocessing_transforms.py:184]: Sounds good. I'm good with doing it. Was just curious.

**Review by emersodb — APPROVED** [review]
Outside of the few small comments I left, this is good to go. Thanks for fixing the unit tests!

**Pull request merged**